src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
Synchronization has been added to transactFully.

Do we know if it's illegal to perform multiple transactions before the previous transaction is read from?  I.e. one of the proposals above was to throw TransactInProgressException but is that a limitation of the protocol of this implementation of the protocol?  If a caller knows how to handle the combined output of multiple transactions, than throwing that exception would restrict the caller from doing something that is perfectly valid.  Going to test this.